omegaconf/dictconfig.py

```diff
@@ -287,29 +288,28 @@ def _validate_and_normalize_key(self, key: Any) -> DictKeyType:
     def _s_validate_and_normalize_key(self, key_type: Any, key: Any) -> DictKeyType:
         if key_type is Any:
             for t in DictKeyType.__args__:  # type: ignore
-                try:
-                    return self._s_validate_and_normalize_key(key_type=t, key=key)
-                except KeyValidationError:
-                    pass
+                if isinstance(key, t):
+                    return cast(DictKeyType, key)
             raise KeyValidationError("Incompatible key type '$KEY_TYPE'")
-        elif key_type == str:
-            if not isinstance(key, str):
-                raise KeyValidationError(
-                    f"Key $KEY ($KEY_TYPE) is incompatible with ({key_type.__name__})"
-                )
-
-            return key
-        elif key_type == int:
-            if not isinstance(key, int):
+        elif key_type == bool and (key == 0 or key == 1):
+            # motivated by an edge case in standard python:
+            # >>> assert hash(0) == hash(False)
+            # >>> assert hash(1) == hash(True)
+            # >>> {True: 123, False: 456}[1]
+            # 123
```

Comment:
Not clear to me whether this is an implementation detail of CPython or an intended Python language feature.

Reply:
Language feature, see this StackOverflow thread that asks exactly this: https://stackoverflow.com/questions/2764017/is-false-0-and-true-1-an-implementation-detail-or-is-it-guaranteed-by-the

Reply:
Oh, very cool, thanks!

Reply:
What I mean is thank you for the information.

Reply:
This special handling is motivated by my [conversation](https://github.com/omry/omegaconf/pull/484#issuecomment-765740259) with Omry regarding consistency between OmegaConf and standard python behavior. In particular, because we have
```
>>> {True: 123}[1]
123
```
in standard python, we want to have
```
>>> DictConfig({True: 123}, key_type=bool)[1]
123
```
in OmegaConf.

Reply:
In all other cases (besides `key_type==bool` and `key in (0,1)`) there is no conversion or coercion of keys, except for some special-cased coersion of `str/int` to `Enum` subclasses.

Reply:
> What I mean is thank you for the information.

I was confused by this and then realized I accidentally clicked the sad emoji. Sorry, I meant to smile! :) (fixed now!)
 
> in OmegaConf, i.e. if `key_type==bool` then we treat integers `0` and `1` as bool.

Yep makes sense. As a side note I find this Python behavior more annoying than useful but I guess it's how it is...